core/src/main/java/org/projectnessie/cel/interpreter/Activation.java

```diff
@@ -21,6 +21,7 @@
 import java.util.function.Function;
 import java.util.function.Supplier;
 import org.projectnessie.cel.common.types.ref.Val;
+import org.projectnessie.cel.shaded.org.antlr.v4.runtime.misc.Pair;
```

Comment:
A pair in generally is fine. Just not such a fan of making the relocated antlr api class part of this API.

WDYT about:
```java
package org.projectnessie.cel.interpreter;

import java.util.Objects;

public interface ResolvedValue {
  Object value();
  boolean present();

  ResolvedValue NULL_VALUE = new NullValue();
  ResolvedValue ABSENT = new AbsentValue();
  static ResolvedValue resolvedValue(Object value) { return new PresentValue(value); }

  final class PresentValue implements ResolvedValue {
    private final Object value;
    private PresentValue(Object value) { this.value = Objects.requireNonNull(value); }
    @Override public Object value() { return value; }
    @Override public boolean present() { return true; }
  }

  final class NullValue implements ResolvedValue {
    private NullValue() {}
    @Override public Object value() { return null; }
    @Override public boolean present() { return true; }
  }

  final class AbsentValue implements ResolvedValue {
    private AbsentValue() {}
    @Override public Object value() { return null; }
    @Override public boolean present() { return false; }
  }}
```

The above can lead to more megamorphic call sites, but OTOH there are already a bunch of these in the code base. Alternatively something like this probably serves better?

```java
public final class ResolvedValue {
  public static final ResolvedValue NULL_VALUE = new ResolvedValue(null, true);
  public static final ResolvedValue ABSENT = new ResolvedValue(null, false);
  public static ResolvedValue resolvedValue(Object value) {
    return new ResolvedValue(Objects.requireNonNull(value), true);
  }

  private final Object value;
  private final boolean present;

  private ResolvedValue(Object value, boolean present) {
    this.value = value;
    this.present = present;
  }

  public Object value() { return value; }
  public boolean present() { return present; }
}
```

Reply:
Your solutions are great, I was just trying to use the existing Class and found that Pair :)

I like your second solution more, to keep things simple. But first one works for me too. WDYT?

Reply:
Yea, the 2nd one is good enough :tm: ;)

Reply:
Not sure (and too sad there's no "null-aware" `Optional`).

A `NULL_VALUE` could mislead other users, because they'd have to do a `if (something != NULL_VALUE)` - so easy to run into issues if they'd do `something != null`, which is different.